### PR TITLE
refactor(headers): Improve docs, fix nits, make formatting faster

### PR DESCRIPTION
**Pull request by pyfisch**
src/header/parsing.rs now uses unsafe get_unchecked() again, I don't
know why it was removed.

[Review on Reviewable](https://reviewable.io/reviews/hyperium/hyper/503)

### CODE REVIEW

File: src/header/parsing.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ pub fn from_one_raw_str<T: str::FromStr>(raw: &[Vec<u8>]) -> Option<T> {
         return None;
     }
     // we JUST checked that raw.len() == 1, so raw[0] WILL exist.
-    if let Ok(s) = str::from_utf8(&raw[0][..]) {
+    if let Ok(s) = str::from_utf8(& unsafe { raw.get_unchecked(0) }[..]) {
```
Comment (reem) [inline src/header/parsing.rs]: LLVM almost certainly elides the bounds check here anyway. I'd personally rather have less unsafe here. I doubt this has a noticeable impact on perf, but I'd be thrilled to be proven wrong.

**Comment by seanmonstar** [comment]
<sup>**[src/header/parsing.rs, line 12 \[r1\]](https://reviewable.io:443/reviews/hyperium/hyper/503#-JoUzxTHZfnIu_OfPQMn-r1-12)** ([raw file](https://github.com/hyperium/hyper/blob/db4d89189a5e200f579cdb1a5d02af44005ea0f9/src/header/parsing.rs#L12)):</sup>
Either way seems fine to me. Or rather, if the "almost certainly" is in fact "absolutely always", then cool. Otherwise, we've done the check above, and a comment reminding anyone about the invariant, so the unsafe looks safe to me as well.

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hyperium/hyper/503)

**Comment by seanmonstar** [comment]
Reviewed files:
- src/header/common/accept.rs @ r1
- src/header/common/access_control_allow_headers.rs @ r1
- src/header/common/access_control_allow_methods.rs @ r1
- src/header/common/access_control_max_age.rs @ r1
- src/header/common/access_control_request_headers.rs @ r1
- src/header/common/access_control_request_method.rs @ r1
- src/header/common/allow.rs @ r1
- src/header/common/authorization.rs @ r1
- src/header/common/cache_control.rs @ r1
- src/header/common/connection.rs @ r1
- src/header/common/content_type.rs @ r1
- src/header/common/cookie.rs @ r1
- src/header/common/etag.rs @ r1
- src/header/common/host.rs @ r1
- src/header/common/if_range.rs @ r1
- src/header/common/last_modified.rs @ r1
- src/header/common/mod.rs @ r1
- src/header/common/pragma.rs @ r1
- src/header/common/set_cookie.rs @ r1
- src/header/internals/item.rs @ r1
- src/header/mod.rs @ r1
- src/header/parsing.rs @ r1
- src/header/shared/charset.rs @ r1
- src/header/shared/encoding.rs @ r1
- src/header/shared/entity.rs @ r1
- src/header/shared/language.rs @ r1

---





Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hyperium/hyper/503)

**Comment by seanmonstar** [comment]
Hrm, seems reviewable is noisier than I was hoping. Oh well, don't worry about it. Just playing with toys, and this was a larger PR, so seemed opportune.

**Pull request merged**